calyx-py/calyx/builder.py

```diff
@@ -0,0 +1,341 @@
+    """Convert a Python object into a control statement.
+
+    This is the machinery for treating shorthand Python expressions as
+    control statements. The rules are:
+
+    * Strings are "enable" leaves.
+    * Groups (and builders) are also enables.
+    * Lists are `seq` statements.
+    * `None` is the `empty` statement.
+    * Otherwise, this should be an actual `Control` value.
+    """
+    if isinstance(obj, ast.Control):
+        return obj
+    elif isinstance(obj, str):
+        return ast.Enable(obj)
+    elif isinstance(obj, ast.Group):
+        return ast.Enable(obj.id.name)
+    elif isinstance(obj, GroupBuilder):
+        return ast.Enable(obj.group.id.name)
+    elif isinstance(obj, list):
+        return ast.SeqComp([as_control(o) for o in obj])
+    elif obj is None:
+        return ast.Empty()
+    else:
+        assert False, f"unsupported control type {type(obj)}"
+
+
+class ControlBuilder:
+    """Wraps control statements for convenient construction."""
+
+    def __init__(self, stmt=None):
+        self.stmt = as_control(stmt)
+
+    def __add__(self, other):
+        """Build sequential composition."""
+        other_stmt = as_control(other)
+
+        if isinstance(self.stmt, ast.Empty):
+            # Special cases for when one side is empty.
+            return ControlBuilder(other_stmt)
+        elif isinstance(other_stmt, ast.Empty):
+            return self
+        elif isinstance(self.stmt, ast.SeqComp) and \
+                isinstance(other_stmt, ast.SeqComp):
+            # Special cases for when we already have at least one seq.
+            return ControlBuilder(
+                ast.SeqComp(self.stmt.stmts + other_stmt.stmts)
+            )
+        elif isinstance(self.stmt, ast.SeqComp):
+            return ControlBuilder(
+                ast.SeqComp(self.stmt.stmts + [other_stmt])
+            )
+        elif isinstance(other_stmt, ast.SeqComp):
+            return ControlBuilder(
+                ast.SeqComp([self.stmt] + other_stmt.stmts)
+            )
+        else:
+            # General case.
+            return ControlBuilder(
+                ast.SeqComp([self.stmt, as_control(other)])
+            )
+
+
+class ExprBuilder:
+    """Wraps an assignment expression.
+
+    This wrapper provides convenient ways to build logical expressions
+    for guards. Use the Python operators &, |, and ~ to build and, or,
+    and not expressions in Calyx.
+
+    It also provides an @ operator to help build conditional
+    expressions, by creating a `CondExprBuilder`.
+    """
+
+    def __init__(self, expr: ast.GuardExpr):
+        self.expr = expr
+
+    def __and__(self, other: 'ExprBuilder'):
+        return ExprBuilder(ast.And(self.expr, other.expr))
+
+    def __or__(self, other: 'ExprBuilder'):
+        return ExprBuilder(ast.Or(self.expr, other.expr))
+
+    def __invert__(self, other: 'ExprBuilder'):
+        return ExprBuilder(ast.Not(self.expr))
+
+    def __matmul__(self, other):
```

Comment:
TIL again :-). For less worthy Python users, maybe explain here or above that this is the `@` symbol? I saw this before reading your comments and was a bit perplexed.

Reply:
Good call; added some docs!